### ISSUE TITLE
Gearmand monitoring

### ISSUE THREAD

**Issue by ilyam8** [netdata/netdata#4817]
We are willing to support monitoring ***[Gearmand](http://gearman.org/)*** with netdata.

To get started, we need a user that will act as a [sponsor](https://docs.netdata.cloud/contributing/#sponsor-a-part-of-netdata) during the implementation.
He/she will assist us with the following:

  - Specify the metrics required and how they should be presented in charts.
  - Specify the alarms that would make sense for each metric.
  - When the implementation passes QA, test the implementation in production.
  - Use the charts and alarms in their day to day work and provide additional feedback after the collector is delivered.

We would also appreciate getting votes :+1: for this issue from the community, to understand the interest in the particular collector.

**Comment by agronick** [comment]
I have made a chart plugin. I think the stacked chart works well for this:

![image](https://user-images.githubusercontent.com/2042303/61063448-f8d76680-a3cd-11e9-9d2b-a050ade75bbd.png)

I will clean it up and post to Github.

**Comment by agronick** [comment]
Here it is https://github.com/agronick/gearman-netdata/tree/master

**Comment by ilyam8** [comment]
Implementation looks a little bit hacky. Are you willing to make a PR btw?

About the gearmand thing, i checked [`datadog` implementation](https://github.com/DataDog/integrations-core/tree/master/gearmand)

They collect summary and per stats queued/idle/active workers statistics. There is no summary stats in your implementation.

I see there can be a lot of tasks. For some reason datadog has hardcoded limitation - [200 tasks](https://github.com/DataDog/integrations-core/blob/7a0b2a7d52593deb95cd927d43dee9da082a8ad5/gearmand/datadog_checks/gearmand/gearmand.py#L18) to collect data from.


```python
            # Display a warning in the info page
            self.warning(
                (
                    "Too many tasks to fetch. "
                    "You must choose the tasks you are interested in by editing the gearmand.yaml configuration file "
                    "or get in touch with Datadog support"
                )
            )
```

:smile:

**Comment by ilyam8** [comment]
The funny thing that 200 is hardcoded and there is no way to change via config file (maybe i am missing something, dunno)

https://github.com/DataDog/integrations-core/blob/7a0b2a7d52593deb95cd927d43dee9da082a8ad5/gearmand/datadog_checks/gearmand/gearmand.py#L71-L80

**Comment by agronick** [comment]
I can make a PR. So you want a summary graph and then the ability for users to set the jobs they care about in the config? Should it default to all jobs if none is set? 

One thing about the gearman stats command is that old jobs never fall off. I don't know if there is a way to remove charts at runtime with Netdata.

One thing I am noticing is that even though I am setting the chart type to stacked, when Netdata loads the page and idle is the only metric not at 0 it makes it a line graph. When it gets another metric it doesn't look as good as the stacked view. It ends up looking like this:

![image](https://user-images.githubusercontent.com/2042303/61067861-a0589700-a3d6-11e9-845a-aaafa3ef8f35.png)

When I refresh the page it goes back to normal:
![image](https://user-images.githubusercontent.com/2042303/61067896-b6feee00-a3d6-11e9-8cf9-97bb9080ef19.png)

Is there a way to force it to show all metrics even if their value is 0 and/or keep it in stacked view? Also is there a way to change colors? The big red blob looks like something is wrong.

**Comment by ilyam8** [comment]
> Is there a way to force it to show all metrics even if their value is 0 and/or keep it in stacked view? 

> show all metrics even if their value is 0

`Dashboard->Settings->Performance->WhichDimensionToShow->All`

> Also is there a way to change colors? The big red blob looks like something is wrong.

No, this is hardcoded. I agree with you, this is wrong. But that is how it is atm.

> the ability for users to set the jobs they care about in the config? Should it default to all jobs if none is set?

dunno, not a gearmand user. What is your opinion on that? Does it makes sense to have summary stats at all? To collect stats for all stats by default?

**Comment by ilyam8** [comment]
> One thing about the gearman stats command is that old jobs never fall off. I don't know if there is a way to remove charts at runtime with Netdata.

Yes, there is a way. 

> I can make a PR.

If you want to work on it please do, but keep in mind it looks like we need a lot of changes :smile: 
If you dont have time/willing this is ok too, i can write it myself, i would do it in `go`.

**Comment by agronick** [comment]
Setting it to show non-zero values has the stacked graphs showing up now. Is there a way to get stacked graphs to show even if idle is the only non-zero value when the chart loads if the user doesn't have that setting set?

I can make it so it shows the overall status for each server and builds charts for specific jobs by setting them in the config file.

I cleaned up the code. What kind of changes were you looking for? If you want to take this over I wouldn't mind. But if you want the help I'm happy to provide it.

**Comment by ilyam8** [comment]
> Is there a way to get stacked graphs to show even if idle is the only non-zero value when the chart loads if the user doesn't have that setting set?

No

> I can make it so it shows the overall status for each server and builds charts for specific jobs by setting them in the config file.

Sounds good to me. 

> I cleaned up the code. What kind of changes were you looking for?

Mainly logic ones. For example you don't need to override `create` method.

> But if you want the help I'm happy to provide it.

Do a PR then, we will continue there. I wont be able to properly review until Sunday/Monday.

**Comment by agronick** [comment]
Ok. How can I set up the charts using data from _get_raw_data() without checking/resetting it on every call?

**Comment by ilyam8** [comment]
Check the [beanstalk collector](https://github.com/netdata/netdata/tree/master/collectors/python.d.plugin/beanstalk), i recall it has some logic (summary stats + per tube stats). It doesnt delete old chart tho, but it is easy to add later.

**Comment by agronick** [comment]
I've updated the plugin to do the following:

- Build a chart showing total statistics (with an option to disable)
- Allow the user to provide a custom list of jobs to build charts for
- Added an option to auto-detect and grab jobs from Gearman on startup. (This is disabled if a list of jobs is provided)

I've tried to make the code look like the other Python modules and I added an example config. If it looks good to you I'll make a pull request.

**Comment by ilyam8** [comment]
@agronick 
i suggest:
 - remove `hide_total` option, we just dont need it, i think it is ok always to show summary chats

About per task charts, there are several options:
 - 1 chart per task, number of chart == number of task
 - simplify init, please do there only simple stuff (initialize attrs)
 - in general flow should be - get tasks; update charts (add chart for new tasks, remove charts for stopped tasks); 
 - use [task_chart_template](https://github.com/netdata/netdata/blob/d768b1292d11bc27c34c50665d234d0e5c848a76/collectors/python.d.plugin/beanstalk/beanstalk.chart.py#L117-L168) func to create new charts (so all charts related stuff will be in once place)

Lets do no add any filtering for now, lets implement:
 - adding charts for new tasks in runtime
 - removing charts for stopped charts in runtime

**Comment by agronick** [comment]
I made the changes you requested. If there is only one worker it ends up showing two identical charts.

**Comment by ilyam8** [comment]
nice, i will open a [new issue in your repo](https://github.com/agronick/gearman-netdata/issues/new) and we continue there

**Comment by ilyam8** [comment]
@agronick i have a second thought about it, please do a PR and i add my suggestion there. This is natural flow :smile: 	

> nice, i will open a new issue in your repo and we will continue there

^^ this makes no sense, sorry for the delay

**Comment by agronick** [comment]
Ok, no problem.

**Comment by ilyam8** [comment]
fixed in #6567
